examples/file-upload.html

```diff
@@ -0,0 +1,370 @@
+   background-color: rgb(160, 0, 0);
+   color: rgb(240, 240, 240);
+  }
+
+  file-upload a.file-upload-clear:active {
+    color: rgba(120, 120, 120, 1);
+    border-bottom: none;
+  }
+
+</style>
+
+<datalist id="file-upload-whitelist">
+  <option value="app-pdf.png"   label="application/pdf">
+  <option value="app-img.png"   label="image/png">
+  <option value="app-img.png"   label="image/jpeg">
+  <option value="txt-plain.png" label="text/plain">
+</datalist>
+
+
+<input multiple id=files name=files[] type=file>
+
+<label for=files ondragover="event.preventDefault()" ondragenter=ondragstart ondragleave=ondragend ondrop=ondragdrop>
+  Click to upload files
+</label>
+
+<a class='file-upload-clear' onclick=onclear>Clear All</a>
+
+<table class="file-upload-list">
+  <thead style="text-align: left">
+    <tr>
+      <th width="1%"></th>
+      <th>Name</th>
+      <th width="10%">Size</th>
+      <th width="1%"></th>
+    </tr>
+  </thead>
+  <tbody>
+    <template name=files>
+      <tr>
+
+        <td>
+          <img src={icon_url} alt={type} hidden={hidden} style="width: 1em; height: 1em">
+        </td>
+        <td>
+          {name}
+        </td>
+        <td>
+          {size}
+        </td>
+        <td class="file-remove-cell">
+          <a onclick=onremove data-index={#}>&times;</a>
+        </td>
+      </tr>
+    </template>
+  </tbody>
+</table>
+
+</template>
+
+<script>
+
+Element `file-upload`
+
+(class extends HTMLElement {
+
+  /**
+   * Initializes the element, setting the files list to []
+   * and generating a file types' whitelist from a datalist
+   * of ID 'file-upload-whitelist'. If no datalist with
+   * such ID is found, it defaults to <code>default_whitelist</code>.
+   *
+   * Also initializes the default values for the label, which
+   * can be changed via <code>set_label</code>.
+   *
+   * @see generate_whitelist
+   * @see default_whitelist
+   * @see set_label
+   */
+  initialize () {
+    this.context.files = []
+
+    this.whitelist = this.generate_whitelist()
+
+    this.default_text = 'Click to upload files'
+    this.on_drag_text = 'Drop here!'
+  }
+
+
+  /**
+   * Listens for changes in the file field, essentially
+   * processing new non-dragged files.
+   *
+   * @param event {Event} The event that has been triggered.
+   */
+  static onchange (event) {
+    if (!!! event.target.files) return
+
+    this.add_files(event.target.files)
+
+    this.render ()
+  }
+
+  /**
+   * Listens for the event of a user removing a file
+   * via the button on the file's entry's row.
+   *
+   * @param event {MouseEvent} The event that has been triggered. 
+   */
+  static onremove (event) {
+    this.context.files.splice(event.target.dataset.index, 1)
+
+    this.render ()
+  }
+
+  /**
+   * Listens for the event of a user clicking on the 'Clear All'
+   * button, and removes all the files accordingly.
+   *
+   * @param event {MouseEvent} The event that has been triggered.
+   */
+  static onclear (event) {
```

Comment:
@VicenteRD your thoughts on using the `<input type=reset>` ? We may just get the `onclear` for free :-) 

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button


![capture d ecran 2017-04-28 a 23 33 45](https://cloud.githubusercontent.com/assets/38223/25552567/54967fce-2c6b-11e7-99c8-80b162e8708f.png)